project/src/main/ui/menu/practice-menu.gd

```diff
@@ -1,147 +1,199 @@
-## levels, so we cache the result.
-func _calculate_lowlights() -> void:
-	if _rank_lowlights:
-		# already calculated
-		return
+## This browses the game files for the region and level corresponding to the player's save data, storing the result
+## in the '_region' and '_level_settings' fields
+func _load_recent_data() -> void:
+	# find the player's previously played non-career region
+	if not _region:
+		for next_region in OtherLevelLibrary.regions:
+			if next_region.id == PlayerData.practice.region_id:
+				_region = next_region
+				break
+	
+	# find the player's previously played career region
+	if not _region:
+		for next_region in CareerLevelLibrary.regions:
+			if next_region.id == PlayerData.practice.region_id:
+				_region = next_region
+				break
 	
-	for difficulty in mode_difficulties[_get_mode()]:
-		var level: LevelSettings = levels["%s %s" % [_get_mode(), difficulty]]
-		_rank_lowlights.append(not PlayerData.level_history.successful_levels.has(level.id))
+	# load the player's previously played level
+	_level_settings.load_from_resource(PlayerData.practice.level_id)
 
 
-func _get_mode() -> String:
-	return $VBoxContainer/Mode.get_selected_mode()
+## Enables/disables focus for inputs as submenus are shown and hidden.
+##
+## When submenus appear over top of the main menu, we temporarily disable focus for all sliders/buttons/etc on the
+## main menu to avoid them from grabbing keyboard focus.
+func _refresh_input_focus_mode() -> void:
+	var submenu_visible: bool = $LevelSubmenu.visible or $RegionSubmenu.visible
+	for node in get_tree().get_nodes_in_group("main_practice_inputs"):
+		node.focus_mode = FOCUS_NONE if submenu_visible else FOCUS_ALL
+
+
+## Updates the level button to show the current level's title.
+func _refresh_level_button() -> void:
+	if _region is CareerRegion:
+		_level_button.text = "%s: %s" % [_region.name, _level_settings.title]
+	else:
+		if _region.has_flag(OtherRegion.FLAG_TRAINING):
+			_level_button.text = _level_settings.title
+		else:
+			_level_button.text = "%s: %s" % [_region.branch_name, _level_settings.title]
+	_level_description_label.text = _level_settings.description
+
+
+## Updates the speed slider with a list of available speeds.
+##
+## For career regions, the player is constrained by the speeds of that particular region.
+##
+## For non-career region, the player can only speed levels up. They cannot slow them down.
+##
+## For rank regions, the player can not make any changes at all.
+func _refresh_speed_selector() -> void:
+	if _region is CareerRegion:
+		# populate speed selector with region speeds
+		var min_index := PieceSpeeds.speed_ids.find(_region.min_piece_speed)
+		var max_index := PieceSpeeds.speed_ids.find(_region.max_piece_speed)
+		var new_speed_names := PieceSpeeds.speed_ids.slice(min_index, max_index)
+		_speed_selector.set_speed_names(new_speed_names)
+		_speed_selector.set_editable(true)
+	else:
+		# populate speed selector with arbitrary speeds
```

Comment:
rephrase `arbitrary speeds` -- maybe something like `a range of speeds specific to that level`